// klayr-commander/dist/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfigFolder = void 0;
const core_1 = require("@oclif/core");
const print_1 = require("./utils/print");
exports.defaultConfigFolder = '.klayr';
const jsonDescription = 'Prints output in JSON format. You can change the default behavior in your config.json file.';
const prettyDescription = 'Prints JSON in pretty format rather than condensed. Has no effect if the output is set to table. You can change the default behavior in your config.json file.';
class BaseCommand extends core_1.Command {
    constructor() {
        super(...arguments);
        this.printFlags = {};
    }
    async finally(error) {
        if (error) {
            this.error(error instanceof Error ? error.message : error);
        }
    }
    async init() {
        const { flags } = await this.parse(this.constructor);
        this.printFlags = flags;
        process.stdout.on('error', (err) => {
            if (err.errno !== 'EPIPE') {
                throw err;
            }
        });
    }
    print(result) {
        (0, print_1.print)({
            ...this.printFlags,
        }).call(this, result);
    }
    printJSON(message) {
        if (typeof message === 'string') {
            this.log(message);
        }
        if (this.printFlags.json) {
            this.log(JSON.stringify(message, undefined, '  '));
        }
        else {
            this.log(JSON.stringify(message));
        }
    }
}
BaseCommand.flags = {
    json: core_1.Flags.boolean({
        char: 'j',
        description: jsonDescription,
        allowNo: true,
    }),
    pretty: core_1.Flags.boolean({
        description: prettyDescription,
        allowNo: true,
    }),
};
exports.default = BaseCommand;
