// klayr-commander/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.plainGeneratorKeysSchema = exports.OWNER_READ_WRITE = exports.RELEASE_URL = exports.DEFAULT_NETWORK = exports.NETWORK = void 0;
var NETWORK;
(function (NETWORK) {
    NETWORK["DEFAULT"] = "default";
    NETWORK["MAINNET"] = "mainnet";
    NETWORK["TESTNET"] = "testnet";
    NETWORK["BETANET"] = "betanet";
    NETWORK["ALPHANET"] = "alphanet";
    NETWORK["DEVNET"] = "devnet";
})(NETWORK = exports.NETWORK || (exports.NETWORK = {}));
exports.DEFAULT_NETWORK = NETWORK.DEFAULT;
exports.RELEASE_URL = 'https://downloads.klayr.xyz/klayr';
exports.OWNER_READ_WRITE = 0o600;
exports.plainGeneratorKeysSchema = {
    $id: '/commander/plainGeneratorKeys',
    type: 'object',
    properties: {
        generatorKey: {
            dataType: 'bytes',
            fieldNumber: 1,
        },
        generatorPrivateKey: {
            dataType: 'bytes',
            fieldNumber: 2,
        },
        blsKey: {
            dataType: 'bytes',
            fieldNumber: 3,
        },
        blsPrivateKey: {
            dataType: 'bytes',
            fieldNumber: 4,
        },
    },
};
