// klayr-framework/dist-node/engine/consensus/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postBlockEventSchema = exports.getHighestCommonBlockResponseSchema = exports.getHighestCommonBlockRequestSchema = exports.getBlocksFromIdResponseSchema = exports.getBlocksFromIdRequestSchema = void 0;
exports.getBlocksFromIdRequestSchema = {
    $id: 'klayr/getBlocksFromIdRequest',
    title: 'Get Blocks From Id Request',
    type: 'object',
    required: ['blockId'],
    properties: {
        blockId: {
            fieldNumber: 1,
            dataType: 'bytes',
            minLength: 32,
            maxLength: 32,
        },
    },
};
exports.getBlocksFromIdResponseSchema = {
    $id: 'klayr/getBlocksFromIdResponse',
    title: 'Get Blocks From Id Response',
    type: 'object',
    required: ['blocks'],
    properties: {
        blocks: {
            type: 'array',
            fieldNumber: 1,
            items: {
                dataType: 'bytes',
            },
        },
    },
};
exports.getHighestCommonBlockRequestSchema = {
    $id: '/klayr/getHighestCommonBlockRequest',
    title: 'Get Highest Common Block Request',
    type: 'object',
    required: ['ids'],
    properties: {
        ids: {
            type: 'array',
            fieldNumber: 1,
            minItems: 1,
            items: {
                dataType: 'bytes',
                minLength: 32,
                maxLength: 32,
            },
        },
    },
};
exports.getHighestCommonBlockResponseSchema = {
    $id: '/klayr/getHighestCommonBlockResponse',
    title: 'Get Highest Common Block Response',
    type: 'object',
    required: ['id'],
    properties: {
        id: {
            dataType: 'bytes',
            fieldNumber: 1,
            minLength: 32,
            maxLength: 32,
        },
    },
};
exports.postBlockEventSchema = {
    $id: '/klayr/postBlockEvent',
    title: 'Post Block Event',
    type: 'object',
    required: ['block'],
    properties: {
        block: {
            dataType: 'bytes',
            fieldNumber: 1,
        },
    },
};
