// klayr-commander/dist/bootstrapping/commands/base_generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseGeneratorCommand = void 0;
const inquirer = require("inquirer");
const flags_1 = require("../../utils/flags");
const base_ipc_client_1 = require("./base_ipc_client");
class BaseGeneratorCommand extends base_ipc_client_1.BaseIPCClientCommand {
    async getPassword(parsedFlags) {
        if (parsedFlags.password) {
            return parsedFlags.password;
        }
        const answers = await inquirer.prompt([
            {
                type: 'password',
                message: 'Enter password to decrypt the encrypted passphrase: ',
                name: 'password',
                mask: '*',
            },
        ]);
        return answers.password;
    }
}
BaseGeneratorCommand.args = [
    {
        name: 'address',
        required: true,
        description: 'Address of an account in a klayr32 format.',
    },
];
BaseGeneratorCommand.flags = {
    ...base_ipc_client_1.BaseIPCClientCommand.flags,
    password: flags_1.flagsWithParser.password,
};
exports.BaseGeneratorCommand = BaseGeneratorCommand;
