// klayr-commander/dist/bootstrapping/commands/base_ipc_client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseIPCClientCommand = void 0;
const apiClient = require("@klayr/api-client");
const core_1 = require("@oclif/core");
const application_1 = require("../../utils/application");
const flags_1 = require("../../utils/flags");
const path_1 = require("../../utils/path");
class BaseIPCClientCommand extends core_1.Command {
    async finally() {
        if (this._client) {
            await this._client.disconnect();
        }
    }
    async init() {
        const { flags } = await this.parse(this.constructor);
        this.baseIPCClientFlags = flags;
        this._dataPath = this.baseIPCClientFlags['data-path']
            ? this.baseIPCClientFlags['data-path']
            : (0, path_1.getDefaultPath)(this.config.pjson.name);
        if (!(0, application_1.isApplicationRunning)(this._dataPath)) {
            throw new Error(`Application at data path ${this._dataPath} is not running.`);
        }
        this._client = await apiClient.createIPCClient(this._dataPath);
        this._schema = this._client.schema;
        this._metadata = this._client.metadata;
    }
    printJSON(message) {
        if (this.baseIPCClientFlags.pretty) {
            this.log(JSON.stringify(message, undefined, '  '));
        }
        else {
            this.log(JSON.stringify(message));
        }
    }
}
BaseIPCClientCommand.flags = {
    'data-path': flags_1.flagsWithParser.dataPath,
    pretty: flags_1.flagsWithParser.pretty,
};
exports.BaseIPCClientCommand = BaseIPCClientCommand;
