// klayr-commander/dist/bootstrapping/commands/hash-onion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HashOnionCommand = void 0;
const core_1 = require("@oclif/core");
const cryptography = require("@klayr/cryptography");
const validator = require("@klayr/validator");
const flags_1 = require("../../utils/flags");
const output_1 = require("../../utils/output");
class HashOnionCommand extends core_1.Command {
    async run() {
        const { flags: { output, count, distance, pretty }, } = await this.parse(HashOnionCommand);
        if (distance <= 0 || !validator.isValidInteger(distance)) {
            throw new Error('Distance flag must be an integer and greater than 0.');
        }
        if (count <= 0 || !validator.isValidInteger(count)) {
            throw new Error('Count flag must be an integer and greater than 0.');
        }
        const seed = cryptography.utils.generateHashOnionSeed();
        const hashBuffers = cryptography.utils.hashOnion(seed, count, distance);
        const hashes = hashBuffers.map(buf => buf.toString('hex'));
        const result = { count, distance, hashes };
        if (output) {
            const res = await (0, output_1.handleOutputFlag)(output, result, 'hash-onion');
            this.log(res);
        }
        else {
            this.printJSON(result, pretty);
        }
    }
    printJSON(message, pretty = false) {
        if (pretty) {
            this.log(JSON.stringify(message, undefined, '  '));
        }
        else {
            this.log(JSON.stringify(message));
        }
    }
}
HashOnionCommand.description = 'Create hash onions to be used by the forger.';
HashOnionCommand.examples = [
    'hash-onion --count=1000000 --distance=2000 --pretty',
    'hash-onion --count=1000000 --distance=2000 --output ~/my_onion.json',
];
HashOnionCommand.flags = {
    output: core_1.Flags.string({
        char: 'o',
        description: 'Output file path',
    }),
    count: core_1.Flags.integer({
        char: 'c',
        description: 'Total number of hashes to produce',
        default: 1000000,
    }),
    distance: core_1.Flags.integer({
        char: 'd',
        description: 'Distance between each hash',
        default: 1000,
    }),
    pretty: flags_1.flagsWithParser.pretty,
};
exports.HashOnionCommand = HashOnionCommand;
