// klayr-commander/dist/bootstrapping/commands/generator/disable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisableCommand = void 0;
const base_generator_1 = require("../base_generator");
class DisableCommand extends base_generator_1.BaseGeneratorCommand {
    async run() {
        const { args, flags } = await this.parse(DisableCommand);
        const { address } = args;
        const password = await this.getPassword(flags);
        if (!this._client) {
            this.error('APIClient is not initialized.');
        }
        await this._client.invoke('generator_updateStatus', {
            address,
            password,
            enable: false,
            height: 0,
            maxHeightGenerated: 0,
            maxHeightPrevoted: 0,
        });
        this.log(`Disabled block generation for ${address}`);
    }
}
DisableCommand.description = 'Disable block generation for given validator address.';
DisableCommand.examples = [
    'generator:disable lskycz7hvr8yfu74bcwxy2n4mopfmjancgdvxq8xz',
    'generator:disable lskycz7hvr8yfu74bcwxy2n4mopfmjancgdvxq8xz --data-path ./data',
    'generator:disable lskycz7hvr8yfu74bcwxy2n4mopfmjancgdvxq8xz --data-path ./data --password your_password',
];
DisableCommand.flags = {
    ...base_generator_1.BaseGeneratorCommand.flags,
};
DisableCommand.args = [...base_generator_1.BaseGeneratorCommand.args];
exports.DisableCommand = DisableCommand;
