// klayr-commander/dist/bootstrapping/commands/generator/export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportCommand = void 0;
const path = require("path");
const flags_1 = require("../../../utils/flags");
const base_ipc_client_1 = require("../base_ipc_client");
const output_1 = require("../../../utils/output");
class ExportCommand extends base_ipc_client_1.BaseIPCClientCommand {
    async run() {
        const { flags } = await this.parse(ExportCommand);
        if (!this._client) {
            this.error('APIClient is not initialized.');
        }
        const allKeys = await this._client.invoke('generator_getAllKeys');
        const statusResponse = await this._client.invoke('generator_getStatus');
        const returnedKeys = allKeys.keys.map(k => {
            if (k.type === 'encrypted') {
                return {
                    address: k.address,
                    encrypted: k.data,
                };
            }
            return {
                address: k.address,
                plain: k.data,
            };
        });
        const returnedGeneratorInfo = statusResponse.status.map(s => ({
            address: s.address,
            height: s.height,
            maxHeightGenerated: s.maxHeightGenerated,
            maxHeightPrevoted: s.maxHeightPrevoted,
        }));
        const output = {
            keys: returnedKeys,
            generatorInfo: returnedGeneratorInfo,
        };
        const filePath = flags.output ? flags.output : path.join(process.cwd(), 'generator_info.json');
        const res = await (0, output_1.handleOutputFlag)(filePath, output, 'generator_info');
        this.log(res);
    }
}
ExportCommand.description = 'Export to <FILE>.';
ExportCommand.examples = [
    'generator:export',
    'generator:export --output /mypath/generator_info.json',
    'generator:export --output /mypath/generator_info.json --data-path ./data ',
];
ExportCommand.flags = {
    ...base_ipc_client_1.BaseIPCClientCommand.flags,
    output: flags_1.flagsWithParser.output,
};
exports.ExportCommand = ExportCommand;
