// klayr-framework/dist-node/state_machine/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContextStoreString = exports.getContextStoreNumber = exports.getContextStoreBool = exports.getContextStoreBigInt = exports.StateMachine = exports.createImmutableMethodContext = exports.createMethodContext = exports.EventQueue = exports.GenesisBlockContext = exports.BlockContext = exports.TransactionContext = exports.NotFoundError = void 0;
const lisk_db_1 = require("@liskhq/lisk-db");
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return lisk_db_1.NotFoundError; } });
var transaction_context_1 = require("./transaction_context");
Object.defineProperty(exports, "TransactionContext", { enumerable: true, get: function () { return transaction_context_1.TransactionContext; } });
var block_context_1 = require("./block_context");
Object.defineProperty(exports, "BlockContext", { enumerable: true, get: function () { return block_context_1.BlockContext; } });
var genesis_block_context_1 = require("./genesis_block_context");
Object.defineProperty(exports, "GenesisBlockContext", { enumerable: true, get: function () { return genesis_block_context_1.GenesisBlockContext; } });
var event_queue_1 = require("./event_queue");
Object.defineProperty(exports, "EventQueue", { enumerable: true, get: function () { return event_queue_1.EventQueue; } });
var method_context_1 = require("./method_context");
Object.defineProperty(exports, "createMethodContext", { enumerable: true, get: function () { return method_context_1.createMethodContext; } });
Object.defineProperty(exports, "createImmutableMethodContext", { enumerable: true, get: function () { return method_context_1.createImmutableMethodContext; } });
__exportStar(require("./types"), exports);
var state_machine_1 = require("./state_machine");
Object.defineProperty(exports, "StateMachine", { enumerable: true, get: function () { return state_machine_1.StateMachine; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getContextStoreBigInt", { enumerable: true, get: function () { return utils_1.getContextStoreBigInt; } });
Object.defineProperty(exports, "getContextStoreBool", { enumerable: true, get: function () { return utils_1.getContextStoreBool; } });
Object.defineProperty(exports, "getContextStoreNumber", { enumerable: true, get: function () { return utils_1.getContextStoreNumber; } });
Object.defineProperty(exports, "getContextStoreString", { enumerable: true, get: function () { return utils_1.getContextStoreString; } });
