// klayr-commander/dist/utils/flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flagsWithParser = exports.flags = void 0;
const core_1 = require("@oclif/core");
const constants_1 = require("../constants");
const messageDescription = `Specifies a source for providing a message to the command. If a string is provided directly as an argument, this option will be ignored. The message must be provided via an argument or via this option. Sources must be one of \`file\` or \`stdin\`. In the case of \`file\`, a corresponding identifier must also be provided.
	Note: if both secret passphrase and message are passed via stdin, the passphrase must be the first line.
	Examples:
	- --message=file:/path/to/my/message.txt
	- --message="hello world"
`;
const passphraseDescription = `Specifies a source for your secret passphrase. Command will prompt you for input if this option is not set.
	Examples:
	- --passphrase='my secret passphrase' (should only be used where security is not important)
`;
const passwordDescription = `Specifies a source for your secret password. Command will prompt you for input if this option is not set.
	Examples:
	- --password=pass:password123 (should only be used where security is not important)
`;
const dataPathDescription = 'Directory path to specify where node data is stored. Environment variable "KLAYR_DATA_PATH" can also be used.';
const offlineDescription = 'Specify whether to connect to a local node or not.';
const networkDescription = 'Default network config to use. Environment variable "KLAYR_NETWORK" can also be used.';
const configDescription = 'File path to a custom config. Environment variable "KLAYR_CONFIG_FILE" can also be used.';
const prettyDescription = 'Prints JSON in pretty format rather than condensed.';
const outputDescription = 'The output directory. Default will set to current working directory.';
const fileDescription = `The file to upload.
	Example:
		--file=./myfile.json
`;
exports.flags = {
    message: {
        char: 'm',
        description: messageDescription,
    },
    passphrase: {
        char: 'p',
        description: passphraseDescription,
    },
    password: {
        char: 'w',
        description: passwordDescription,
    },
    dataPath: {
        char: 'd',
        description: dataPathDescription,
    },
    offline: {
        description: offlineDescription,
    },
    network: {
        char: 'n',
        description: networkDescription,
    },
    config: {
        char: 'c',
        description: configDescription,
    },
    pretty: {
        description: prettyDescription,
    },
    output: {
        char: 'o',
        description: outputDescription,
    },
    json: {
        char: 'j',
        description: 'Print the transaction in JSON format.',
    },
    senderPublicKey: {
        char: 's',
        description: "Set a custom senderPublicKey property for the transaction, to be used when account address does not correspond to signer's private key",
    },
    file: {
        char: 'f',
        description: fileDescription,
    },
};
exports.flagsWithParser = {
    dataPath: core_1.Flags.string({
        ...exports.flags.dataPath,
        env: 'KLAYR_DATA_PATH',
    }),
    network: core_1.Flags.string({
        ...exports.flags.network,
        env: 'KLAYR_NETWORK',
        default: constants_1.DEFAULT_NETWORK,
    }),
    config: core_1.Flags.string({
        ...exports.flags.config,
        env: 'KLAYR_CONFIG_FILE',
    }),
    pretty: core_1.Flags.boolean(exports.flags.pretty),
    passphrase: core_1.Flags.string(exports.flags.passphrase),
    output: core_1.Flags.string({
        ...exports.flags.output,
        default: process.cwd(),
    }),
    password: core_1.Flags.string(exports.flags.password),
    offline: core_1.Flags.boolean({
        ...exports.flags.offline,
    }),
    json: core_1.Flags.boolean(exports.flags.json),
    senderPublicKey: core_1.Flags.string(exports.flags.senderPublicKey),
    chainID: core_1.Flags.string(exports.flags.chainID),
    file: core_1.Flags.string(exports.flags.file),
};
