// klayr-commander/dist/utils/path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPidPath = exports.getForgerDBPath = exports.getModuleDBPath = exports.getStateDBPath = exports.getBlockchainDBPath = exports.ensureConfigDir = exports.getConfigDirs = exports.getNetworkConfigFilesPath = exports.getConfigFilesPath = exports.getFullPath = exports.getDefaultPath = void 0;
const path = require("path");
const fs = require("fs-extra");
const os = require("os");
const defaultDir = '.klayr';
const getConfigPath = (dataPath) => path.join(dataPath, 'config');
const getDefaultPath = (name) => path.join(os.homedir(), defaultDir, name);
exports.getDefaultPath = getDefaultPath;
const getFullPath = (dataPath) => path.resolve(dataPath);
exports.getFullPath = getFullPath;
const getConfigFilesPath = (dataPath, configDirIncluded = false) => {
    const basePath = configDirIncluded ? path.join(dataPath) : path.join(dataPath, 'config');
    return {
        basePath,
        genesisBlockFilePath: path.join(basePath, 'genesis_block.blob'),
        configFilePath: path.join(basePath, 'config.json'),
    };
};
exports.getConfigFilesPath = getConfigFilesPath;
const getNetworkConfigFilesPath = (dataPath, network, configDirIncluded = false) => {
    const basePath = configDirIncluded
        ? path.join(dataPath, network)
        : path.join(dataPath, 'config', network);
    return {
        basePath,
        genesisBlockFilePath: path.join(basePath, 'genesis_block.blob'),
        configFilePath: path.join(basePath, 'config.json'),
    };
};
exports.getNetworkConfigFilesPath = getNetworkConfigFilesPath;
const getConfigDirs = (dataPath, configDirIncluded = false) => {
    const configPath = configDirIncluded ? dataPath : getConfigPath(dataPath);
    fs.ensureDirSync(configPath);
    const files = fs.readdirSync(configPath);
    return files.filter(file => fs.statSync(path.join(configPath, file)).isDirectory());
};
exports.getConfigDirs = getConfigDirs;
const ensureConfigDir = (dataPath) => fs.ensureDirSync(path.join(dataPath, 'config'));
exports.ensureConfigDir = ensureConfigDir;
const getBlockchainDBPath = (dataPath) => path.join(dataPath, 'data', 'blockchain.db');
exports.getBlockchainDBPath = getBlockchainDBPath;
const getStateDBPath = (dataPath) => path.join(dataPath, 'data', 'state.db');
exports.getStateDBPath = getStateDBPath;
const getModuleDBPath = (dataPath) => path.join(dataPath, 'data', 'module.db');
exports.getModuleDBPath = getModuleDBPath;
const getForgerDBPath = (dataPath) => path.join(dataPath, 'data', 'forger.db');
exports.getForgerDBPath = getForgerDBPath;
const getPidPath = (dataPath) => path.join(dataPath, 'tmp', 'pids', 'controller.pid');
exports.getPidPath = getPidPath;
