// klayr-commander/dist/bootstrapping/commands/keys/export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportCommand = void 0;
const flags_1 = require("../../../utils/flags");
const base_ipc_client_1 = require("../base_ipc_client");
const output_1 = require("../../../utils/output");
class ExportCommand extends base_ipc_client_1.BaseIPCClientCommand {
    async run() {
        const { flags } = await this.parse(ExportCommand);
        if (!this._client) {
            this.error('APIClient is not initialized.');
        }
        const response = await this._client.invoke('generator_getAllKeys');
        const keys = response === null || response === void 0 ? void 0 : response.keys.map(k => {
            if (k.type === 'encrypted') {
                return {
                    address: k.address,
                    encrypted: k.data,
                };
            }
            return {
                address: k.address,
                plain: k.data,
            };
        });
        if (flags.output) {
            const res = await (0, output_1.handleOutputFlag)(flags.output, { keys }, 'keys');
            this.log(res);
        }
    }
}
ExportCommand.description = 'Export to <FILE>.';
ExportCommand.examples = [
    'keys:export --output /mypath/keys.json',
    'keys:export --output /mypath/keys.json --data-path ./data ',
];
ExportCommand.flags = {
    ...base_ipc_client_1.BaseIPCClientCommand.flags,
    output: {
        ...flags_1.flagsWithParser.output,
    },
};
exports.ExportCommand = ExportCommand;
