// klayr-commander/dist/bootstrapping/commands/keys/import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportCommand = void 0;
const core_1 = require("@oclif/core");
const fs = require("fs-extra");
const base_ipc_client_1 = require("../base_ipc_client");
class ImportCommand extends base_ipc_client_1.BaseIPCClientCommand {
    async run() {
        const { flags } = await this.parse(ImportCommand);
        if (!this._client) {
            this.error('APIClient is not initialized.');
        }
        const fileData = fs.readJSONSync(flags['file-path']);
        const keys = fileData.keys.map(k => {
            let type;
            let returnData;
            if (k.encrypted && Object.keys(k.encrypted).length !== 0) {
                type = 'encrypted';
                returnData = k.encrypted;
            }
            else {
                type = 'plain';
                returnData = k.plain;
            }
            return {
                address: k.address,
                type,
                data: returnData,
            };
        });
        for (const key of keys) {
            await this._client.invoke('generator_setKeys', key);
        }
    }
}
ImportCommand.description = 'Import from <FILE>.';
ImportCommand.examples = [
    'keys:import --file-path ./my/path/keys.json',
    'keys:import --file-path ./my/path/keys.json --data-path ./data ',
];
ImportCommand.flags = {
    ...base_ipc_client_1.BaseIPCClientCommand.flags,
    'file-path': core_1.Flags.string({
        char: 'f',
        description: 'Path of the file to import from',
        required: true,
    }),
};
exports.ImportCommand = ImportCommand;
