// klayr-commander/dist/commands/generate/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const base_bootstrap_command_1 = require("../../base_bootstrap_command");
class PluginCommand extends base_bootstrap_command_1.default {
    async run() {
        var _a;
        const { args: { name }, flags: { standalone, output, registry }, } = await this.parse(PluginCommand);
        const regexWhitespace = /\s/g;
        const regexCamelCase = /[a-z]+((\d)|([A-Z0-9][a-z0-9]+))*([A-Z])?/;
        const regexAlphabets = /[^A-Za-z]/;
        if (!regexCamelCase.test(name) || regexWhitespace.test(name) || regexAlphabets.test(name)) {
            this.error('Invalid plugin name');
        }
        if (standalone) {
            return this._runBootstrapCommand('klayr:init:plugin', {
                name,
                projectPath: (_a = output !== null && output !== void 0 ? output : process.env.INIT_CWD) !== null && _a !== void 0 ? _a : process.cwd(),
                registry,
            });
        }
        if (!this._isKlayrappDir(process.cwd())) {
            this.error('You can run this command only in klayr app directory. Run "klayr init --help" command for more details.');
        }
        return this._runBootstrapCommand('klayr:generate:plugin', {
            name,
        });
    }
}
PluginCommand.description = 'Creates custom plugin.';
PluginCommand.examples = [
    'generate:plugin myPlugin',
    'generate:plugin myPlugin --standalone --output ./my_plugin',
];
PluginCommand.args = [
    {
        name: 'name',
        description: 'Name of the plugin.',
        required: true,
    },
];
PluginCommand.flags = {
    ...base_bootstrap_command_1.default.flags,
    standalone: core_1.Flags.boolean({
        description: 'Create a standalone plugin package.',
    }),
    output: core_1.Flags.string({
        description: 'Path to create the plugin.',
        char: 'o',
        dependsOn: ['standalone'],
    }),
    registry: core_1.Flags.string({
        description: 'URL of a registry to download dependencies from.',
        dependsOn: ['standalone'],
    }),
};
exports.default = PluginCommand;
