// klayr-commander/dist/bootstrapping/templates/klayr-template-ts/generators/command_generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const ts_morph_1 = require("ts-morph");
const Generator = require("yeoman-generator");
const convert_1 = require("../../../../utils/convert");
class CommandGenerator extends Generator {
    constructor(args, opts) {
        super(args, opts);
        this._moduleName = opts.moduleName;
        this._commandName = opts.commandName;
        this._moduleFileName = (0, convert_1.camelToSnake)(this._moduleName);
        this._templatePath = (0, path_1.join)(__dirname, '..', 'templates', 'command');
        this._commandClass = `${(0, convert_1.camelToPascal)(this._commandName)}Command`;
        this._commandFileName = `${(0, convert_1.camelToSnake)(this._commandName)}_command`;
        this._moduleClass = `${(0, convert_1.camelToPascal)(this._moduleName)}Module`;
    }
    writing() {
        this.fs.copyTpl(`${this._templatePath}/src/app/modules/commands/command.ts`, this.destinationPath(`src/app/modules/${this._moduleFileName}/commands/${this._commandFileName}.ts`), {
            moduleName: this._moduleName,
            commandName: this._commandName,
            commandClass: this._commandClass,
        }, {}, { globOptions: { dot: true, ignore: ['.DS_Store'] } });
        this.fs.copyTpl(`${this._templatePath}/test/unit/modules/commands/command.spec.ts`, this.destinationPath(`test/unit/modules/${this._moduleFileName}/commands/${this._commandFileName}.spec.ts`), {
            moduleName: this._moduleName,
            commandName: this._commandName,
            commandFileName: this._commandFileName,
            commandClass: this._commandClass,
        }, {}, { globOptions: { dot: true, ignore: ['.DS_Store'] } });
    }
    async registerAsset() {
        this.log('Registering command...');
        const project = new ts_morph_1.Project();
        project.addSourceFilesAtPaths('src/app/**/*.ts');
        const moduleFile = project.getSourceFileOrThrow(`src/app/modules/${this._moduleFileName}/module.ts`);
        moduleFile.addImportDeclaration({
            namedImports: [this._commandClass],
            moduleSpecifier: `./commands/${this._commandFileName}`,
        });
        const moduleClass = moduleFile.getClassOrThrow(this._moduleClass);
        const property = moduleClass.getInstancePropertyOrThrow('commands');
        const value = property.getStructure().initializer;
        if (value === '[]' || value === '') {
            property.set({ initializer: `[new ${this._commandClass}(this.stores, this.events)]` });
        }
        else if (value.endsWith(']')) {
            property.set({
                initializer: `${value.slice(0, -1)}, new ${this._commandClass}(this.stores, this.events)]`,
            });
        }
        else {
            this.log('Asset can not be registered. Please register it by yourself.');
        }
        moduleFile.organizeImports();
        await moduleFile.save();
    }
}
exports.default = CommandGenerator;
